File: packages/core/src/semantics/glob.ts
/**
 * Tiny fnmatch-style glob matcher, sufficient for tag-name patterns
 * (`*-rc*`, `v[0-9]*`, etc). We avoid pulling in `minimatch` because the
 * surface we need is very small and `core` aims to stay dependency-free.
 *
 * Supported metacharacters: `*` (any run), `?` (single char), `[...]` (class).
 */
export function matchesGlob(name: string, pattern: string): boolean {
  return globToRegex(pattern).test(name);
}

const cache = new Map<string, RegExp>();

function globToRegex(pattern: string): RegExp {
  const cached = cache.get(pattern);
  if (cached) return cached;

  let re = '^';
  for (let i = 0; i < pattern.length; i++) {
    const ch = pattern[i]!;
    if (ch === '*') {
      re += '.*';
    } else if (ch === '?') {
      re += '.';
    } else if (ch === '[') {
      const close = pattern.indexOf(']', i + 1);
      if (close === -1) {
        re += '\\[';
      } else {
        re += '[' + pattern.slice(i + 1, close).replace(/\\/g, '\\\\') + ']';
        i = close;
      }
    } else if (/[\\^$.+|(){}]/.test(ch)) {
      re += '\\' + ch;
    } else {
      re += ch;
    }
  }
  re += '$';
  const compiled = new RegExp(re);
  cache.set(pattern, compiled);
  return compiled;
}

File: packages/jetbrains-plugin/src/main/kotlin/com/shiplens/jetbrains/core/semantics/Glob.kt
package com.shiplens.jetbrains.core.semantics

import java.util.concurrent.ConcurrentHashMap
import java.util.regex.Pattern

/**
 * Tiny fnmatch-style glob matcher, mirror of @shiplens/core's `semantics/glob.ts`.
 * Supports: `*` (any run), `?` (single char), `[...]` (character class).
 */
internal object Glob {
  private val cache = ConcurrentHashMap<String, Pattern>()

  fun matches(name: String, pattern: String): Boolean = compile(pattern).matcher(name).matches()

  private fun compile(pattern: String): Pattern = cache.computeIfAbsent(pattern) { translate(it) }

  private fun translate(pattern: String): Pattern {
    val sb = StringBuilder("^")
    var i = 0
    while (i < pattern.length) {
      when (val ch = pattern[i]) {
        '*' -> sb.append(".*")
        '?' -> sb.append('.')
        '[' -> {
          val close = pattern.indexOf(']', i + 1)
          if (close == -1) {
            sb.append("\\[")
          } else {
            sb.append('[').append(pattern.substring(i + 1, close).replace("\\", "\\\\")).append(']')
            i = close
          }
        }

        '\\', '^', '$', '.', '+', '|', '(', ')', '{', '}' -> sb.append('\\').append(ch)
        else -> sb.append(ch)
      }
      i++
    }
    sb.append('$')
    return Pattern.compile(sb.toString())
  }
}
